# Week 1 — Checkpoint 1: Data Loading & Preview
### Dataset: `house_sale.csv` (bina.az — real estate sale listings from Baku and other cities)


## 1. Importing the required libraries

At this stage I only import the core libraries needed to explore the structure of the data (`pandas`, `numpy`). Visualization and modeling libraries will be needed in later checkpoints, so I'm not adding them yet.

In [14]:
import pandas as pd
import numpy as np

pd.set_option('display.max_columns', 60)
pd.set_option('display.width', 150)


## 2. Loading the dataset

https://www.kaggle.com/datasets/sehriyarmemmedli/binaaz-sale-project

In [15]:
#!/bin/bash
!curl -L -o binaaz-sale-project.zip\
  https://www.kaggle.com/api/v1/datasets/download/sehriyarmemmedli/binaaz-sale-project


  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
100 39.8M  100 39.8M    0     0  56.3M      0 --:--:-- --:--:-- --:--:--  120M


In [16]:
!unzip /content/binaaz-sale-project.zip

Archive:  /content/binaaz-sale-project.zip
replace house_sale.csv? [y]es, [n]o, [A]ll, [N]one, [r]ename: y
  inflating: house_sale.csv          


In [17]:
df = pd.read_csv("house_sale.csv")

In [18]:
df

,id_x,rel_url,estate_rel_url_x,datetime_scrape_x,price,currency_x,location,attributes,city_when,city,day_x,hour_x,repair,vip,featured,products_label,bill_of_sale,mortgage,img_url,id_y,estate_id,estate_rel_url_y,datetime_scrape_y,description,unit_price,total_price,currency_y,owner_name,owner_title,shop_name,shop_title,address,lat,lng,updated,views,day_y,hour_y,estate_details_id_x,Binanın növü,Kateqoriya,Mərtəbə,Otaq sayı,Sahə,Torpaq sahəsi,Təmir,Çıxarış,İpoteka,estate_details_id_y,estate_rel_url,extra_info
0,5df36281-6dc6-4d5d-89a7-5fcfa86f7608,/alqi-satqi?page=174,/items/4521724,2024-10-05 22:07:37.60613+00,499999.0,AZN,Səbail r.,"4 otaqlı, 145 m², 7/9 mərtəbə","Bakı, dünən 23:52",bakı,05.10.2024,23:52,Təmirli,vipped,featured,NaN,Çıxarış var,NaN,https://bina.azstatic.com/uploads/f460x345/202...,92e82ea2-e1f3-4c2d-a284-151efd99281e,5df36281-6dc6-4d5d-89a7-5fcfa86f7608,/items/4521724,2024-10-05 22:14:10.089116+00,"Səbail Rayonu, İzzət Nəbiyev küçəsi, Fəxri Xiy...",3 450 AZN/m²,499999.0,AZN,Kamran,mülkiyyətçi,NaN,NaN,İzzət Nəbiyev küç.,40.358817,49.824092,yeniləndi: dünən 23:52,1155,05.10.2024,23:52,92e82ea2-e1f3-4c2d-a284-151efd99281e,NaN,Köhnə tikili,7 / 9,4.0,145 m²,NaN,var,var,NaN,92e82ea2-e1f3-4c2d-a284-151efd99281e,/items/4521724,Şəhidlər xiyabanı * Dağüstü parkı * Səbail r.
1,883e20f0-8872-49a5-8b4b-63b8301b5f8f,/alqi-satqi?page=250,/items/4669294,2024-10-05 22:07:37.60613+00,77000.0,AZN,Biləcəri q.,"4 otaqlı, 90 m²","Bakı, dünən 23:56",bakı,05.10.2024,23:56,Təmirli,NaN,NaN,NaN,NaN,NaN,https://bina.azstatic.com/uploads/f460x345/202...,505eaf81-6bc8-4094-9b00-aa82066548ee,883e20f0-8872-49a5-8b4b-63b8301b5f8f,/items/4669294,2024-10-05 22:14:10.089116+00,"Biləcəridə Abidəyə yaxin 91,92,202 saylı marşr...",NaN,77000.0,AZN,Dasinmaz Emlak,vasitəçi (agent),NaN,NaN,Biləcəri qəs.,40.420897,49.807035,yeniləndi: 04 oktyabr 2024,218,04.10.2024,NaN,505eaf81-6bc8-4094-9b00-aa82066548ee,NaN,Həyət evi/Bağ evi,NaN,4.0,90 m²,1.3 sot,var,yoxdur,NaN,505eaf81-6bc8-4094-9b00-aa82066548ee,/items/4669294,Binəqədi r.* Biləcəri q.
2,55c36fb1-a3af-476e-ba17-a81f6795be8d,/alqi-satqi?page=250,/items/4669293,2024-10-05 22:07:37.60613+00,92000.0,AZN,İnşaatçılar m.,"3 otaqlı, 60 m²","Bakı, dünən 23:55",bakı,05.10.2024,23:55,Təmirli,NaN,NaN,NaN,Çıxarış var,NaN,https://bina.azstatic.com/uploads/f460x345/202...,fa63b201-999d-43b5-a61a-778d9d79a6c6,55c36fb1-a3af-476e-ba17-a81f6795be8d,/items/4669293,2024-10-05 22:14:10.089116+00,Salam əleykum. \nİnşaatçılar metrosuna yaxın m...,NaN,92000.0,AZN,Məhəmməd,vasitəçi (agent),NaN,NaN,Mirzə Cabbar Məmmədzadə küç.,40.390293,49.802656,yeniləndi: 04 oktyabr 2024,190,04.10.2024,NaN,fa63b201-999d-43b5-a61a-778d9d79a6c6,NaN,Həyət evi/Bağ evi,NaN,3.0,60 m²,0.1 sot,var,var,NaN,fa63b201-999d-43b5-a61a-778d9d79a6c6,/items/4669293,İnşaatçılar m.* Yasamal r.
3,acf1aa8d-a46a-40f5-b6f2-f7a449569337,/alqi-satqi?page=250,/items/4647811,2024-10-05 22:07:37.60613+00,95000.0,AZN,Qaraçuxur q.,130 m²,"Bakı, dünən 23:55",bakı,05.10.2024,23:55,Təmirli,vipped,featured,NaN,Çıxarış var,İpoteka var,https://bina.azstatic.com/uploads/f460x345/202...,5db56980-05cc-4925-b55b-f58fb3f4d2b6,acf1aa8d-a46a-40f5-b6f2-f7a449569337,/items/4647811,2024-10-05 22:14:10.089116+00,Barter maraqlidir üstünlük maşina verilir\nHər...,NaN,95000.0,AZN,Elçin,mülkiyyətçi,NaN,NaN,Qaraçuxur qəs.,40.393614,49.981553,yeniləndi: 04 oktyabr 2024,1314,04.10.2024,NaN,5db56980-05cc-4925-b55b-f58fb3f4d2b6,NaN,Obyekt,NaN,NaN,130 m²,NaN,var,var,var,5db56980-05cc-4925-b55b-f58fb3f4d2b6,/items/4647811,Suraxanı r.* Qaraçuxur q.
4,22d840df-9283-4112-bc71-7432511fc776,/alqi-satqi?page=250,/items/4638863,2024-10-05 22:07:37.60613+00,220000.0,AZN,Əhmədli m.,"3 otaqlı, 100 m², 15/16 mərtəbə","Bakı, dünən 23:52",bakı,05.10.2024,23:52,Təmirli,NaN,NaN,Agentlik,Çıxarış var,NaN,https://bina.azstatic.com/uploads/f460x345/202...,d71cf9a9-86dd-4162-b540-6252a8659a09,22d840df-9283-4112-bc71-7432511fc776,/items/4638863,2024-10-05 22:14:10.089116+00,Əhmədli qəs. Qaçaq Nəbi küçəsi 3 otaga duzelm...

## 3. General overview (shape, head)

First, I check how many rows and columns the dataset has, and look at the first few rows to get a general sense of the data — what fields exist, what the values look like, etc.

In [19]:
print("Number of rows:", df.shape[0])
print("Number of columns:", df.shape[1])


Number of rows: 100775
Number of columns: 51


In [20]:
df.head()


,id_x,rel_url,estate_rel_url_x,datetime_scrape_x,price,currency_x,location,attributes,city_when,city,day_x,hour_x,repair,vip,featured,products_label,bill_of_sale,mortgage,img_url,id_y,estate_id,estate_rel_url_y,datetime_scrape_y,description,unit_price,total_price,currency_y,owner_name,owner_title,shop_name,shop_title,address,lat,lng,updated,views,day_y,hour_y,estate_details_id_x,Binanın növü,Kateqoriya,Mərtəbə,Otaq sayı,Sahə,Torpaq sahəsi,Təmir,Çıxarış,İpoteka,estate_details_id_y,estate_rel_url,extra_info
0,5df36281-6dc6-4d5d-89a7-5fcfa86f7608,/alqi-satqi?page=174,/items/4521724,2024-10-05 22:07:37.60613+00,499999.0,AZN,Səbail r.,"4 otaqlı, 145 m², 7/9 mərtəbə","Bakı, dünən 23:52",bakı,05.10.2024,23:52,Təmirli,vipped,featured,NaN,Çıxarış var,NaN,https://bina.azstatic.com/uploads/f460x345/202...,92e82ea2-e1f3-4c2d-a284-151efd99281e,5df36281-6dc6-4d5d-89a7-5fcfa86f7608,/items/4521724,2024-10-05 22:14:10.089116+00,"Səbail Rayonu, İzzət Nəbiyev küçəsi, Fəxri Xiy...",3 450 AZN/m²,499999.0,AZN,Kamran,mülkiyyətçi,NaN,NaN,İzzət Nəbiyev küç.,40.358817,49.824092,yeniləndi: dünən 23:52,1155,05.10.2024,23:52,92e82ea2-e1f3-4c2d-a284-151efd99281e,NaN,Köhnə tikili,7 / 9,4.0,145 m²,NaN,var,var,NaN,92e82ea2-e1f3-4c2d-a284-151efd99281e,/items/4521724,Şəhidlər xiyabanı * Dağüstü parkı * Səbail r.
1,883e20f0-8872-49a5-8b4b-63b8301b5f8f,/alqi-satqi?page=250,/items/4669294,2024-10-05 22:07:37.60613+00,77000.0,AZN,Biləcəri q.,"4 otaqlı, 90 m²","Bakı, dünən 23:56",bakı,05.10.2024,23:56,Təmirli,NaN,NaN,NaN,NaN,NaN,https://bina.azstatic.com/uploads/f460x345/202...,505eaf81-6bc8-4094-9b00-aa82066548ee,883e20f0-8872-49a5-8b4b-63b8301b5f8f,/items/4669294,2024-10-05 22:14:10.089116+00,"Biləcəridə Abidəyə yaxin 91,92,202 saylı marşr...",NaN,77000.0,AZN,Dasinmaz Emlak,vasitəçi (agent),NaN,NaN,Biləcəri qəs.,40.420897,49.807035,yeniləndi: 04 oktyabr 2024,218,04.10.2024,NaN,505eaf81-6bc8-4094-9b00-aa82066548ee,NaN,Həyət evi/Bağ evi,NaN,4.0,90 m²,1.3 sot,var,yoxdur,NaN,505eaf81-6bc8-4094-9b00-aa82066548ee,/items/4669294,Binəqədi r.* Biləcəri q.
2,55c36fb1-a3af-476e-ba17-a81f6795be8d,/alqi-satqi?page=250,/items/4669293,2024-10-05 22:07:37.60613+00,92000.0,AZN,İnşaatçılar m.,"3 otaqlı, 60 m²","Bakı, dünən 23:55",bakı,05.10.2024,23:55,Təmirli,NaN,NaN,NaN,Çıxarış var,NaN,https://bina.azstatic.com/uploads/f460x345/202...,fa63b201-999d-43b5-a61a-778d9d79a6c6,55c36fb1-a3af-476e-ba17-a81f6795be8d,/items/4669293,2024-10-05 22:14:10.089116+00,Salam əleykum. \nİnşaatçılar metrosuna yaxın m...,NaN,92000.0,AZN,Məhəmməd,vasitəçi (agent),NaN,NaN,Mirzə Cabbar Məmmədzadə küç.,40.390293,49.802656,yeniləndi: 04 oktyabr 2024,190,04.10.2024,NaN,fa63b201-999d-43b5-a61a-778d9d79a6c6,NaN,Həyət evi/Bağ evi,NaN,3.0,60 m²,0.1 sot,var,var,NaN,fa63b201-999d-43b5-a61a-778d9d79a6c6,/items/4669293,İnşaatçılar m.* Yasamal r.
3,acf1aa8d-a46a-40f5-b6f2-f7a449569337,/alqi-satqi?page=250,/items/4647811,2024-10-05 22:07:37.60613+00,95000.0,AZN,Qaraçuxur q.,130 m²,"Bakı, dünən 23:55",bakı,05.10.2024,23:55,Təmirli,vipped,featured,NaN,Çıxarış var,İpoteka var,https://bina.azstatic.com/uploads/f460x345/202...,5db56980-05cc-4925-b55b-f58fb3f4d2b6,acf1aa8d-a46a-40f5-b6f2-f7a449569337,/items/4647811,2024-10-05 22:14:10.089116+00,Barter maraqlidir üstünlük maşina verilir\nHər...,NaN,95000.0,AZN,Elçin,mülkiyyətçi,NaN,NaN,Qaraçuxur qəs.,40.393614,49.981553,yeniləndi: 04 oktyabr 2024,1314,04.10.2024,NaN,5db56980-05cc-4925-b55b-f58fb3f4d2b6,NaN,Obyekt,NaN,NaN,130 m²,NaN,var,var,var,5db56980-05cc-4925-b55b-f58fb3f4d2b6,/items/4647811,Suraxanı r.* Qaraçuxur q.
4,22d840df-9283-4112-bc71-7432511fc776,/alqi-satqi?page=250,/items/4638863,2024-10-05 22:07:37.60613+00,220000.0,AZN,Əhmədli m.,"3 otaqlı, 100 m², 15/16 mərtəbə","Bakı, dünən 23:52",bakı,05.10.2024,23:52,Təmirli,NaN,NaN,Agentlik,Çıxarış var,NaN,https://bina.azstatic.com/uploads/f460x345/202...,d71cf9a9-86dd-4162-b540-6252a8659a09,22d840df-9283-4112-bc71-7432511fc776,/items/4638863,2024-10-05 22:14:10.089116+00,Əhmədli qəs. Qaçaq Nəbi küçəsi 3 otaga duzelm...

In [21]:
df.sample(5, random_state=42)


,id_x,rel_url,estate_rel_url_x,datetime_scrape_x,price,currency_x,location,attributes,city_when,city,day_x,hour_x,repair,vip,featured,products_label,bill_of_sale,mortgage,img_url,id_y,estate_id,estate_rel_url_y,datetime_scrape_y,description,unit_price,total_price,currency_y,owner_name,owner_title,shop_name,shop_title,address,lat,lng,updated,views,day_y,hour_y,estate_details_id_x,Binanın növü,Kateqoriya,Mərtəbə,Otaq sayı,Sahə,Torpaq sahəsi,Təmir,Çıxarış,İpoteka,estate_details_id_y,estate_rel_url,extra_info
980,8794a55a-4493-42e0-8f40-f2c7792bff47,/alqi-satqi?page=294,/items/4645538,2024-10-05 22:07:37.60613+00,75000.0,AZN,Səbail r.,"1 otaqlı, 40 m², 1/5 mərtəbə","Bakı, dünən 17:31",bakı,05.10.2024,17:31,Təmirli,NaN,NaN,NaN,NaN,NaN,https://bina.azstatic.com/uploads/f460x345/202...,d16c3888-b05c-45d4-8272-734adcabd264,8794a55a-4493-42e0-8f40-f2c7792bff47,/items/4645538,2024-10-05 22:14:10.089116+00,"Badamdar qəsəbəsində, 5 mərtəbəli Apartmant ti...",1 880 AZN/m²,75000.0,AZN,Maqsud,vasitəçi (agent),NaN,NaN,Badamdar qəs.,40.339561,49.809115,yeniləndi: 04 oktyabr 2024,368,04.10.2024,NaN,d16c3888-b05c-45d4-8272-734adcabd264,NaN,Yeni tikili,1 / 5,1.0,40 m²,NaN,var,yoxdur,NaN,d16c3888-b05c-45d4-8272-734adcabd264,/items/4645538,Səbail r.
87688,bd4381e5-9d53-4409-b8bd-17e71dad13d3,/alqi-satqi?page=346,/items/4471739,2024-11-09 20:45:46.632914,510000.0,AZN,Nardaran q.,"3 otaqlı, 119 m², 4/5 mərtəbə","Bakı, dünən 12:00",bakı,09.11.2024,12:00,Təmirli,NaN,NaN,Agentlik,NaN,NaN,https://bina.azstatic.com/uploads/f460x345/202...,6510aaf7-a61f-494c-ba65-7ef0d89bb796,bd4381e5-9d53-4409-b8bd-17e71dad13d3,/items/4471739,2024-11-09 20:50:03.404512,SEA BREEZE RESORT\nDel Ventus\nПроект Palazzo ...,4 290 AZN/m²,510000.0,AZN,Azerbaijan Realty and Partners,vasitəçi (agent),Azerbaijan Realty and Partners,Daşınmaz əmlak agentliyi,Nardaran qəs.,40.591405,49.985218,yeniləndi: 08 noyabr 2024,801,08.11.2024,NaN,6510aaf7-a61f-494c-ba65-7ef0d89bb796,NaN,Yeni tikili,4 / 5,3.0,119 m²,NaN,var,yoxdur,NaN,6510aaf7-a61f-494c-ba65-7ef0d89bb796,/items/4471739,Sea Breeze * Sabunçu r.* Nardaran q.
46608,9139800d-ac6d-4d3d-a515-3b486932357f,/alqi-satqi?page=360,/items/4620526,2024-10-18 23:50:50.528821+00,440000.0,AZN,Şah İsmayıl Xətai m.,"3 otaqlı, 124 m², 3/7 mərtəbə","Bakı, dünən 12:59",bakı,18.10.2024,12:59,NaN,NaN,NaN,Agentlik,NaN,NaN,https://bina.azstatic.com/uploads/f460x345/202...,aac5209e-1743-4dea-b658-5c98bc31727f,9139800d-ac6d-4d3d-a515-3b486932357f,/items/4620526,2024-10-18 23:56:27.858464+00,"️ Xətai,Ağ Şəhər\n📍 Fəvvarələr küç\n💬 124m²,3/...",3 550 AZN/m²,440000.0,AZN,"""Megapolis Estate"" İlqar Məmmədov",vasitəçi (agent),Megapolis Estate,Daşınmaz əmlak agentliyi,Fəvvarələr küçəsi,40.380477,49.880395,yeniləndi: 17 oktyabr 2024,274,17.10.2024,NaN,aac5209e-1743-4dea-b658-5c98bc31727f,NaN,Yeni tikili,3 / 7,3.0,124 m²,NaN,yoxdur,yoxdur,NaN,aac5209e-1743-4dea-b658-5c98bc31727f,/items/4620526,ASAN Xidmət №2 * Ağ şəhər * Şah İsmayıl Xətai ...
92171,937c288a-2394-4c79-a1ba-4f62ef92c741,/alqi-satqi?page=362,/items/4755548,2024-11-12 20:43:10.396293,110000.0,AZN,Suraxanı r.,"2 otaqlı, 52 m², 1/17 mərtəbə","Bakı, dünən 23:06",bakı,12.11.2024,23:06,Təmirli,NaN,NaN,NaN,Çıxarış var,İpoteka var,https://bina.azstatic.com/uploads/f460x345/202...,33c95a8b-c3e3-4e5f-a2d4-c7954d4b61da,937c288a-2394-4c79-a1ba-4f62ef92c741,/items/4755548,2024-11-12 20:46:36.658788,Coox tecili!!!!\nSatılır 2 otaqlı yeni tikili ...,2 120 AZN/m²,110000.0,AZN,Zulya x.,vasitəçi (agent),NaN,NaN,A. Fətullayev küç.,40.371260,49.977152,yeniləndi: dünən 23:06,79,12.11.2024,23:06,33c95a8b-c3e3-4e5f-a2d4-c7954d4b61da,NaN,Yeni tikili,1 / 17,2.0,52 m²,NaN,var,var,var,33c95a8b-c3e3-4e5f-a2d4-c7954d4b61da,/items/4755548,Suraxanı r.
45698,b282e16d-d13d-4500-8a7f-5d2d75b55394,/alqi-satqi?page=319,/items/4449952,2024-10-18 23:50:50.528821+00,430000.0,AZN,Mərdəkan q.,"4 otaqlı, 220 m²","Bakı, dünən 16:29",bakı,18.10.2024,16:29,Təmirli,NaN,NaN,Agentlik,Çıxarış var,NaN,https://bina.azstatic.com/uploads/f460x345/202...,

**Observation:** The dataset consists of 100,775 rows and 51 columns. Some column names are in English (e.g. `price`, `location`, `city`), while others are in Azerbaijani (e.g. `Otaq sayı` = number of rooms, `Mərtəbə` = floor, `Təmir` = renovation) — because the dataset was scraped from the bina.az website and the site's own filter field names ended up directly as column names. I'll take this into account during the cleaning stage (Checkpoint 2).

## 4. Column data types (dtypes)

I check how each column is recognized by Python/pandas. This matters because some columns that should logically be numeric might be read as `object` (text), or date fields might remain as text — I'll fix these in the upcoming checkpoints.

In [22]:
df.dtypes


,0
id_x,object
rel_url,object
estate_rel_url_x,object
datetime_scrape_x,object
price,float64
currency_x,object
location,object
attributes,object
city_when,object
city,object


In [23]:
df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100775 entries, 0 to 100774
Data columns (total 51 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   id_x                 100775 non-null  object 
 1   rel_url              100775 non-null  object 
 2   estate_rel_url_x     100775 non-null  object 
 3   datetime_scrape_x    100775 non-null  object 
 4   price                100775 non-null  float64
 5   currency_x           100775 non-null  object 
 6   location             100775 non-null  object 
 7   attributes           100775 non-null  object 
 8   city_when            100775 non-null  object 
 9   city                 100775 non-null  object 
 10  day_x                100775 non-null  object 
 11  hour_x               100775 non-null  object 
 12  repair               81612 non-null   object 
 13  vip                  8596 non-null    object 
 14  featured             3149 non-null    object 
 15  products_label   

**Observation:**
- `price` and `total_price` are already read correctly as `float64` — calculations can be performed directly on the price columns.
- `lat`, `lng` (coordinates) are `float64`, and `views` is `int64` — correctly typed.
- `Otaq sayı` (number of rooms) is `float64` (because it contains NaNs, pandas keeps it as float even though it should logically be an integer).
- Date/time columns (`datetime_scrape_x`, `datetime_scrape_y`, `updated`, `day_x`, `hour_x`, etc.) are still in `object` (text) format — converting these to proper `datetime` type will be handled in Checkpoint 2.
- The remaining columns (`location`, `city`, `attributes`, `description`, `address`, etc.) are categorical/text in nature and remain `object`, which is expected.


## 5. Checking for null values

I calculate how many null values exist in each column and what percentage that represents of the total row count. Looking at the percentage matters, because for example "2,000 missing values" may look negligible in a 100,000-row dataset but could be critical in a dataset with only 3,000 rows.

In [24]:
null_counts = df.isnull().sum()
null_percent = (null_counts / len(df) * 100).round(2)

null_summary = pd.DataFrame({
    "null_count": null_counts,
    "null_percent": null_percent
}).sort_values("null_count", ascending=False)

null_summary[null_summary["null_count"] > 0]


,null_count,null_percent
Binanın növü,100621,99.85
featured,97626,96.88
vip,92179,91.47
Torpaq sahəsi,85204,84.55
hour_y,85144,84.49
İpoteka,67836,67.31
mortgage,67836,67.31
shop_title,28757,28.54
shop_name,28757,28.54
products_label,28145,27.93


**Observation (initial — just noting it here, the actual handling will happen in Checkpoint 2):**

- `Binanın növü` (building type) is almost entirely empty (99.8%) — practically unusable in its current form, a candidate for dropping.
- `featured` (96.9%), `vip` (91.5%), `Torpaq sahəsi` / land area (84.6%), `hour_y` (84.5%) — very high missing rates. Some of these (`vip`, `featured`) are likely boolean-style fields — meaning a missing value probably means "this status does not apply" (MNAR — systematic missingness), not something to simply drop.
- `mortgage`/`İpoteka` (67%), `shop_name`/`shop_title` (28.5%), `products_label` (27.9%), `unit_price`/`Mərtəbə` (floor) (24.6%), `bill_of_sale` (21%) — moderate missing rates; the reason behind these also needs investigation (it could be related to whether the listing was posted by an individual or a real estate agency).
- `Otaq sayı` (number of rooms) (9.3%), `repair`/`Təmir` (renovation) (19% / 6%) — relatively low, but still columns that need to be filled carefully.
- `owner_name`/`owner_title` (0.65%), `description` (0.26%), `unit_price` — very little missingness.

The "why is this missing" question for each of these columns will be investigated in detail in Checkpoint 2 (null-handling strategy) — instead of simply dropping them, I'll examine whether the missingness is random (MCAR) or systematic (MNAR).

## 6. Checking for duplicate rows

First, I check for rows that are fully identical across all columns. Then I check the uniqueness of the identifier columns (`estate_id`, `id_x`, `id_y`) — because even without full-row duplicates, the same listing could still have been recorded more than once at different scrape times.

In [25]:
print("All-column duplicate row count:", df.duplicated().sum())

All-column duplicate row count: 0


In [26]:
print("Total row count:", len(df))
print("Unique estate_id count:", df["estate_id"].nunique())
print("Unique id_x count:", df["id_x"].nunique())
print("Unique id_y count:", df["id_y"].nunique())
print("Unique rel_url count:", df["rel_url"].nunique(), "(this is the scraped page URL, not a listing ID — being low is expected)")


Total row count: 100775
Unique estate_id count: 100775
Unique id_x count: 100775
Unique id_y count: 100775
Unique rel_url count: 411 (this is the scraped page URL, not a listing ID — being low is expected)


**Observation:**
- Full-row duplicates: **0** — there are no rows in the dataset that are identical across every column.
- `estate_id`, `id_x`, and `id_y` each have 100,775 unique values, matching the total row count — meaning every row represents a distinct property listing, with no duplication at the ID level.
- `rel_url` has only 411 unique values — this is because the column represents the **page** the listing was scraped from during pagination (e.g. `/alqi-satqi?page=174`), not the individual listing itself. This is not a duplication issue; it simply reflects what the column represents.
- An additional point I noticed: the dataset contains paired columns like `id_x`/`id_y`, `estate_rel_url_x`/`estate_rel_url_y`, `estate_details_id_x`/`estate_details_id_y` — this likely comes from merging two separate tables (a listing table + a listing-details table). I'll examine more closely in Checkpoint 2 whether these columns simply duplicate each other's information.


## Checkpoint 1 — Summary

- The dataset consists of **100,775 rows and 51 columns** (real estate sale listings from Baku and other cities, scraped from bina.az).
- Most column types were read correctly (`price`, `total_price`, `lat`, `lng`, `views` — numeric), but the date/time columns are still in text (`object`) format and need conversion.
- Null values are present in more than 20 columns, ranging from very low (~0.3%) to nearly completely empty (~99.8%); many appear to be systematic (MNAR) in nature (e.g. `vip`, `featured`, `Torpaq sahəsi`).
- No full-row duplicates exist, and the unique ID columns (`estate_id`, `id_x`, `id_y`) match the total row count — no duplicate-listing issue is apparent.
- There are signs that the dataset was merged from two separate sources (listing + details), indicated by the `_x`/`_y` column pairs — this will be examined carefully in the next stage.
